README.md
<img src = "resources/logo.png" width = "50%" alt = "Logo"/>

# InterSystems-HiThere-GIFGenerator

Generates a [InterSystems](https://www.intersystems.com/) HiThere Banner for your profiles.

![Example](out/Example.gif)

---

-   [Requirements](#requirements)
-   [Usage](#usage)
-   [Configuration](#configuration)
-   [Custom Mode](#custom-mode)
-   [Bugs](#bugs)
-   [Release Notes](#release-notes)

---

## Requirements

-   [Python](https://www.python.org/) 3.7 or higher
-   [Pillow](https://pypi.org/project/Pillow/) 9.2.0 or higher
-   Download the _[DINAlternate-Bold.ttf](https://www.cufonfonts.com/font/din-alternate)_ and add it to the directory

---

## Usage

1.  Download the [latest release](https://github.com/phil1436/InterSystemsHiThereGifGenerator/releases/latest) and extract it to a folder of your choice.
2.  Open a terminal and navigate to the folder.
3.  Run the following command:

```bash
python3 Main.py
```

---

## Configuration

### Help

To get a list of all available properties, run the following command:

```bash
python3 Main.py -help
```

### Duration

The duration of the GIF can be configured with the `duration` property, followed by the duration, where a lower value means a faster animation. Default is 80.

_Run with a duration of 100:_

```bash
python3 Main.py -duration 100
```

> **Note:** The duration is not the same as the duration of the GIF. The duration is the time between each frame, while the GIF duration is the time between the first and the last frame.

### Hashtag

Add the hashtag _**#ItIsNotJustAJob**_ to the end with the `hashtag` property.

```bash
python3 Main.py -hashtag
```

>

### Hold

How long your name will be displayed can be configured with the `hold` property, followed by the number of frames. Default is 15.

_Will hold the name for 20 frames:_

```bash
python3 Main.py -hold 20
```

## Font

Change the font with the `font` property, followed by the path to the font. Default is `DINAlternate-Bold.ttf`.

```bash
python3 Main.py -font path/to/font.ttf
```

### Output

Change the output folder with the `output` property, followed by the path to the folder. Default is `out`.

```bash
python3 Main.py -output path/to/folder
```

### Whitespaces

Change the whitespaces with the `whitespaces` property, followed by `on` or `off`. Default is `on`.

```bash
python3 Main.py -whitespaces off
```

### Run in custom mode

You can run the application in custom mode, where you can specify. This is done with the `custom` property, the program will then ask you for the text you want to display.

```bash
python3 Main.py -custom
```

---

## Custom Mode

To customize the text, you want to display, you can run the application in custom mode.

```bash
InterSystemsHiThereGenerator> python3 Main.py -custom -whitespaces on

      *******************************************
      **** InterSystems-HiThere-GIFGenerator ****
      *******************************************

***Custom Mode***
Part 1
Enter first text (empty to stop): Part
Enter second text (can be empty): One
Part 2
Enter first text (empty to stop): Part
Enter second text (can be empty): Two
Part 3
Enter first text (empty to stop): Part
Enter second text (can be empty): Three
Part 4
Enter first text (empty to stop):
GIF Generated!
You can find it here: out/PartOne.gif

by phil1436
```

This will generate the following GIF:

![CustomModeExample](resources/CustomModeExample.gif)

> Tip: For full customization, you can also use the _[HiThereGIFGenerator.py](HiThereGIFGenerator.py)_ file directly. This file contains the `HTGIFGenerator` class, which can be used to generate the GIF. You can find an example in the _[Main.py](Main.py)_ file.

---

## Bugs

-   _no known bugs_

---

## [Release Notes](https://github.com/phil1436/InterSystemsHiThereGifGenerator/blob/master/CHANGELOG.md)

### [v0.0.2](https://github.com/phil1436/InterSystemsHiThereGifGenerator/tree/0.0.2)

### [v0.0.1](https://github.com/phil1436/InterSystemsHiThereGifGenerator/tree/0.0.1)

---

by Philipp B.

_This application is **not** supported by InterSystems Corporation._


CHANGELOG.md
# Change Log _InterSystems-HiThere-GIFGenerator_

---

## [v0.0.2](https://github.com/phil1436/InterSystemsHiThereGifGenerator/tree/0.0.2) (2023-2-10)

-   You can set the font with the `font` property.
-   You can add the hashtag _**#ItIsNotJustAJob**_ with the `hashtag` property.

---

## [v0.0.1](https://github.com/phil1436/InterSystemsHiThereGifGenerator/tree/0.0.1) (2023-2-9)

-   _Initial release_

---


HiThereGIFGenerator.py
from PIL import Image, ImageSequence
from PIL import ImageDraw
from PIL import ImageFont


class HTGIFPart:
    def __init__(self, path_to_font_ttf: str, first: str, second: str = None):
        self.first = first
        self.second = second
        self.first_color = (0, 181, 175)
        self.second_color = (51, 54, 149)
        self.font = path_to_font_ttf

    # change the color of the first part
    def set_first_color(self, color):
        self.first_color = color

    # change the color of the second part
    def set_second_color(self, color):
        self.second_color = color

    # toString
    def to_string(self):
        if (self.second == None):
            return self.first
        return self.first + self.second


class HTGIFGenerator:
    def __init__(self, path_to_background_image: str, file_output: str = None, start_gif: str = None, end_gif: str = None):
        self.duration = 80
        self.hold = 15
        self.background = path_to_background_image
        width, height = Image.open(path_to_background_image).size
        self.max_width = width
        self.max_height = height
        self.start_gif = start_gif
        self.end_gif = end_gif
        if (file_output == None):
            self.file_output = "HiThere.gif"
        else:
            self.file_output = file_output
        self.parts = []
        self.default_font = None

    # change the duration of the gif
    def set_duration(self, duration: int):
        self.duration = duration

    # change the hold time
    def set_hold(self, hold: int):
        self.hold = hold

    # change the output file
    def set_file_output(self, file_output_path: str):
        self.file_output = file_output_path

    # change the start gif
    def set_start_gif(self, start_gif_path: str):
        self.start_gif = start_gif_path

    # change the end gif
    def set_end_gif(self, end_gif_path: str):
        self.end_gif = end_gif_path

    # change the max width of the text
    def set_max_width(self, max_width: int):
        self.max_width = max_width

    # change the max height of the text
    def set_max_height(self, max_height: int):
        self.max_height = max_height

    # add a part to the gif
    def add_part(self, part: HTGIFPart):
        self.parts.append(part)

    # add a part to the gif with default colors and the default font
    def add_default_part(self, first: str, second: str = None, font: str = None):
        if font == None:
            font = self.default_font
            if font == None:
                raise ValueError(
                    "You should set a default font or pass a font to the function!")
        self.add_part(HTGIFPart(font, first, second))

    # set and change the default font
    def set_default_font(self, path_to_font_ttf: str):
        self.default_font = path_to_font_ttf

    # reset the parts
    def reset(self):
        self.parts = []

    # generate the gif
    def generate(self):
        if (self.parts == []):
            raise ValueError("There should be at least one part!")

        frames = []

        # add start gif
        if (self.start_gif != None):
            for frame in ImageSequence.Iterator(Image.open(self.start_gif)):
                frame = frame.copy()
                frames.append(frame)

        # add the parts
        for parts in self.parts:
            part_imgs = self.generate_images(parts)

            # add start part
            for img in part_imgs:
                frames.append(img)
            # add hold
            for i in range(self.hold):
                frames.append(part_imgs[-1])
            # add end part
            for img in part_imgs[::-1]:
                frames.append(img)
            frames.append(Image.open(self.background))
            frames.append(Image.open(self.background))

        # add start gif
        if (self.end_gif != None):
            for frame in ImageSequence.Iterator(Image.open(self.end_gif)):
                frame = frame.copy()
                frames.append(frame)

        frame_one = frames[0]
        frame_one.save(self.file_output, format="GIF", append_images=frames[1:],
                       save_all=True, duration=self.duration, loop=0, disposal=2)

    # generate the images for a part
    def generate_images(self, part: HTGIFPart):

        img = Image.open(self.background)
        I1 = ImageDraw.Draw(img)

        # adjust font size
        myFont = ImageFont.truetype(part.font, 1)
        width, height = img.size
        size_width = I1.textlength(part.to_string(), myFont)
        size_height = I1.textbbox((0, 0), part.to_string(), font=myFont)[3]
        pt = 1
        # get the maximal font size
        while size_width < self.max_width - 30 and size_height < self.max_height - 30:
            pt += 1
            myFont = ImageFont.truetype(part.font, pt)
            size_width = I1.textlength(part.to_string(), myFont)
            size_height = I1.textbbox((0, 0), part.to_string(), font=myFont)[3]

        images = []

        size_width = I1.textlength(part.to_string(), myFont)

        y = height / 2 - \
            (I1.textbbox((0, 0), part.to_string(), font=myFont)[3] / 2)

        first = ""
        for char in part.first:
            img = Image.open(self.background)
            first += char
            I1 = ImageDraw.Draw(img)

            letters_width = I1.textlength(first, myFont)

            x = width/2 - (letters_width / 2)

            I1.text((x, y), first,
                    fill=part.first_color, font=myFont)

            images.append(img)

        first_width = I1.textlength(first, myFont)

        if (part.second == None):
            return images

        second = ""
        for char in part.second:
            img = Image.open(self.background)
            second += char
            I1 = ImageDraw.Draw(img)

            # get the center of the image
            width, height = img.size

            letters_width = I1.textlength(first + second, myFont)
            x = width/2 - (letters_width / 2)

            I1.text((x, y), first,
                    fill=part.first_color, font=myFont)

            I1.text((x + first_width, y), second,
                    fill=part.second_color, font=myFont)
            images.append(img)
        return images


Main.py
import sys
from HiThereGIFGenerator import HTGIFGenerator, HTGIFPart

VERSION = "0.0.2"

print("""
      *******************************************
      **** InterSystems-HiThere-GIFGenerator ****
      *******************************************
      """)

template_file = "imgs/template.png"
start_gif = "imgs/start.gif"
end_gif = "imgs/end.gif"
file_output_path = "out"
duration = 80
hold = 15
max_width = 900
max_height = 270
font = 'DINAlternate-Bold.ttf'
run_custom_mode = False
whitespace = " "
add_hashtag = False

# check for arguments
for i in range(len(sys.argv)):
    arg = sys.argv[i].lower()
    if (arg == "-v" or arg == "-version"):
        print("Version: " + VERSION)
        print("")
        print("by phil1436")
        quit()
    if (arg == "-help"):
        print("""
              
    Configurations:
    
    [-v|-version] - show version
    [-h|-hold] [int] - set the hold time
    [-d|-duration] [int] - set the duration time
    [-o|-output] [path] - set the output path
    [-f|-font] [path_to_ttf_file] - set the font, by referencing a ttf file
    [-c|-custom] - run in custom mode
    [-w|-whitespaces|-whitespace] [on|off] - set if whitespaces should be used
    [-hashtag|-itisnotjustajob] - adds the hashtag #ItIsNotJustAJob to the end of the gif
              """)
        print("")
        print("by phil1436")
        quit()
    if (arg == "-d" or arg == "-duration"):
        duration = int(sys.argv[i+1])
        print("Duration: " + str(duration))
    if (arg == "-h" or arg == "-hold"):
        hold = int(sys.argv[i+1])
        print("Hold: " + str(hold) + " frames")
    if (arg == "-o" or arg == "-output"):
        file_output_path = sys.argv[i+1]
        print("Output: " + file_output_path)
    if (arg == "-c" or arg == "-custom"):
        run_custom_mode = True
        print("***Custom Mode***")
    if (arg == "-w" or arg == "-whitespaces" or arg == "-whitespace"):
        if (sys.argv[i+1].lower() == "off"):
            whitespace = ""
            print("Whitespaces: off")
        else:
            print("Whitespaces: on")
    if (arg == "-hashtag" or arg == "-itisnotjustajob"):
        add_hashtag = True
    if (arg == "-f" or arg == "-font"):
        font = sys.argv[i+1]


# generate gif
generator = HTGIFGenerator(path_to_background_image=template_file,
                           start_gif=start_gif,
                           end_gif=end_gif)
generator.set_duration(duration)
generator.set_hold(hold)
generator.set_max_width(max_width)
generator.set_max_height(max_height)
generator.set_default_font(font)

if not run_custom_mode:
    # get name from user
    name = input("Enter your name: ")

    # check if name is beween 2 and 10 characters
    while (len(name) < 2 or len(name) > 15):
        print("Please enter a name between 2 and 15 characters")
        name = input("Enter your name: ")
    # filepaths
    fp_out = file_output_path+"/HiThereIm"+name+".gif"

    generator.set_file_output(fp_out)

    generator.add_default_part("Hi" + whitespace, "There!")
    generator.add_default_part("I'm" + whitespace,  name)
else:
    counter = 1
    fp_out = file_output_path+"/"
    while (True):
        print("Part " + str(counter))
        first = input("Enter first text (empty to stop): ")
        if first == "":
            break
        second = input("Enter second text (can be empty): ")

        if counter == 1:
            fp_out = fp_out + first + second + ".gif"
            generator.set_file_output(fp_out)

        if second == "":
            second = None
        generator.add_default_part(first + whitespace, second)
        counter += 1

if (add_hashtag):
    generator.add_default_part("#ItIsNot", "JustAJob")


generator.generate()

print("GIF Generated!")
print("You can find it here: " + fp_out)
print("")
print("by phil1436")
